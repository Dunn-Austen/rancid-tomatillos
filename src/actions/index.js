export const addMovies = movies => ({
  type: 'ADD_MOVIES',
  movies: movies
});

export const updateUser = user => ({
  type: 'UPDATE_USER',
  user: user
});

export const updateLoggedInStatus = status => ({
  type: 'UPDATE_LOGGEDIN_STATUS',
  status: status
});

export const handleError = errorMessage => ({
  type: 'HANDLE_ERROR',
  errorMessage
});

export const isLoading = loadingStatus => ({
  //Passing in a boolean, where loadingStatus key is assigned: true/false
  type: 'IS_LOADING',
  loadingStatus
});
