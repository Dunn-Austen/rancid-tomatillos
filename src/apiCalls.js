export const fetchAllMovies = () => {
  return fetch(
    'https://rancid-tomatillos.herokuapp.com/api/v1/movies'
  ).then(response => response.json());
};

const body = {email: "marge@turing.io", password: "password123"};
// We will need to make the body based on the this.state.email and
// this.state.password from the local state of the form.  This is just set up to
// be able to easily merge with form when that component is ready.

const options = {
  method: 'POST',
  body: JSON.stringify(body),
  headers: {
    'Content-Type': 'application/json'
  }
}
const fetchUserLogin = (options) => {
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options)
    .then(response => response.json())
    // The .then()s below here won't stay in this file, we will chain them on
    // wherever we invoke this fetch function
    // .then(data => console.log(`Number ${data.user.id} is ${data.user.name}`))
    .then(data => {
      // here we would want to send the data.user.id and data.user.name to the store
    })
}
fetchUserLogin(options);

const mapDispatchToProps = dispatch => ({
  updateUser: user => dispatch(updateUser(user))
});

// will need to add the about mapDispatchToProps to the FORM to submit the USER
// when it gets the user obj back
