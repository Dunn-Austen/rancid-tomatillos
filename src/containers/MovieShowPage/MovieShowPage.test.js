import React from 'react';
import { shallow } from 'enzyme';
import { MovieShowPage, mapStateToProps, mapDispatchToProps }
  from './MovieShowPage';
import { updateUser, handleError } from '../../actions';
import { updateRatings, fetchRatings, deleteRating } from '../../apiCalls';

jest.mock('../../apiCalls.js');

describe('MovieShowPage', () => {
  let wrapper, mockUpdateUser, mockHandleError;

  beforeEach(() => {
    mockUpdateUser = jest.fn();
    mockHandleError = jest.fn();
    wrapper = shallow(<MovieShowPage
      id={2}
      title="Ad Astra"
      poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
      release_date="2019-09-17"
      overview="The near future, humanity to look to the stars an....."
      average_rating={5.142857142857143}
      isLoggedIn={true}
      updateUser={mockUpdateUser}
      handleError={mockHandleError}
      user={ {id: 9, name: 'Marge', email: 'marge@turing.io', ratings: [
        {id:45,
          user_id:9,
          movie_id:2,
          rating:8,
          created_at:"2019-12-25T20:16:34.893Z",
          updated_at:"2019-12-25T20:16:34.893Z"},
        {id:46,
          user_id:9,
          movie_id:10,
          rating:10,
          created_at:"2019-12-25T20:30:21.606Z",
          updated_at:"2019-12-25T20:30:21.606Z"}
        ]} }
    />);
    deleteRating.mockImplementation(() => {
      return Promise.resolve()
    });
    updateRatings.mockImplementation(() => {
      return Promise.resolve({ rating: { user_id: 9, movie_id: 2, rating: 2 } })
    });
    fetchRatings.mockImplementation(() => {
      return Promise.resolve({ ratings: [
        {id: 45, user_id: 9, movie_id: 8, rating: 8,
          created_at: "2019-12-25T20:16:34.893Z",
          updated_at: "2019-12-25T20:16:34.893Z"
        },
        {id: 46, user_id: 9, movie_id: 10, rating: 5,
          created_at: "2019-12-25T20:30:21.606Z",
          updated_at: "2019-12-25T20:30:21.606Z"
        } ] })
    })
  })

  it('should invoke updateUser when fetchRatings resolves after handleRatingsUpdates is invoked', async () => {
    mockUpdateUser = jest.fn();
    mockHandleError = jest.fn();
    wrapper = shallow(<MovieShowPage
      id={2}
      title="Ad Astra"
      poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
      release_date="2019-09-17"
      overview="The near future, humanity to look to the stars an....."
      average_rating={5.142857142857143}
      isLoggedIn={true}
      updateUser={mockUpdateUser}
      handleError={mockHandleError}
      user={ {id: 9, name: 'Marge', email: 'marge@turing.io', ratings: [
        {id:45,
          user_id:9,
          movie_id:3,
          rating:8,
          created_at:"2019-12-25T20:16:34.893Z",
          updated_at:"2019-12-25T20:16:34.893Z"},
        {id:46,
          user_id:9,
          movie_id:10,
          rating:10,
          created_at:"2019-12-25T20:30:21.606Z",
          updated_at:"2019-12-25T20:30:21.606Z"}
        ]} }
    />);
    let mockEvent = { target: {value: '2'} };
    let mockNewRatings = {id: 9, name: 'Marge', email: 'marge@turing.io',
        ratings: [
          {id:45,
            user_id:9,
            movie_id:3,
            rating:8,
            created_at:"2019-12-25T20:16:34.893Z",
            updated_at:"2019-12-25T20:16:34.893Z"},
          {id:46,
            user_id:9,
            movie_id:10,
            rating:10,
            created_at:"2019-12-25T20:30:21.606Z",
            updated_at:"2019-12-25T20:30:21.606Z"},
          {id:47,
            user_id:9,
            movie_id:2,
            rating:2,
            created_at:"2019-12-30T20:30:21.606Z",
            updated_at:"2019-12-30T20:30:21.606Z"} ]};
    updateRatings.mockImplementation(() => {
      return Promise.resolve({ rating: { user_id: 9, movie_id: 2, rating: 2 } })
    });
    fetchRatings.mockImplementation(() => {
      return Promise.resolve({ ratings: [
        {id:45,
          user_id:9,
          movie_id:3,
          rating:8,
          created_at:"2019-12-25T20:16:34.893Z",
          updated_at:"2019-12-25T20:16:34.893Z"},
        {id:46,
          user_id:9,
          movie_id:10,
          rating:10,
          created_at:"2019-12-25T20:30:21.606Z",
          updated_at:"2019-12-25T20:30:21.606Z"},
        {id:47,
          user_id:9,
          movie_id:2,
          rating:2,
          created_at:"2019-12-30T20:30:21.606Z",
          updated_at:"2019-12-30T20:30:21.606Z"}
         ] })
    })
    await wrapper.instance().handleRatingsUpdates(mockEvent);
    await wrapper.instance().forceUpdate();
    expect(mockUpdateUser).toHaveBeenCalledWith(mockNewRatings);
  });

  it('should match the MovieShowPage Snapshot with no logged in user', () => {
    wrapper = shallow(<MovieShowPage
      id={2}
      title="Ad Astra"
      poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
      release_date="2019-09-17"
      overview="The near future, humanity to look to the stars an....."
      average_rating={5.142857142857143}
      isLoggedIn={false}
      updateUser={mockUpdateUser}
      handleError={mockHandleError}
      user={ {} }
      />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot when user is logged in and rating is less than 10',
    () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot when user is logged in and rating is 10',
    () => {
      wrapper = shallow(<MovieShowPage
        id={2}
        title="Ad Astra"
        poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
        backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
        release_date="2019-09-17"
        overview="The near future, humanity to look to the stars an....."
        average_rating={5.142857142857143}
        isLoggedIn={true}
        updateUser={mockUpdateUser}
        handleError={mockHandleError}
        user={ {id: 9, name: 'Marge', email: 'marge@turing.io', ratings: [
          {id:45,
            user_id:9,
            movie_id:10,
            rating:8,
            created_at:"2019-12-25T20:16:34.893Z",
            updated_at:"2019-12-25T20:16:34.893Z"},
          {id:46,
            user_id:9,
            movie_id:2,
            rating:10,
            created_at:"2019-12-25T20:30:21.606Z",
            updated_at:"2019-12-25T20:30:21.606Z"}
          ]} }
      />)
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot when logged in and no user rating yet', () => {
    wrapper = shallow(<MovieShowPage
      id={2}
      title="Ad Astra"
      poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
      release_date="2019-09-17"
      overview="The near future, humanity to look to the stars an....."
      average_rating={5.142857142857143}
      isLoggedIn={true}
      updateUser={mockUpdateUser}
      handleError={mockHandleError}
      user={ {id: 9, name: 'Marge', email: 'marge@turing.io', ratings: []} }
    />)
    expect(wrapper).toMatchSnapshot();
  });

  it('should invoke deleteRating on click', () => {
    wrapper = shallow(<MovieShowPage
      id={2}
      title="Ad Astra"
      poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
      release_date="2019-09-17"
      overview="The near future, humanity to look to the stars an....."
      average_rating={5.142857142857143}
      isLoggedIn={true}
      updateUser={mockUpdateUser}
      handleError={mockHandleError}
      user={ {id: 9, name: 'Marge', email: 'marge@turing.io', ratings: [
        {id:45,
          user_id:9,
          movie_id:2,
          rating:8,
          created_at:"2019-12-25T20:16:34.893Z",
          updated_at:"2019-12-25T20:16:34.893Z"},
        {id:46,
          user_id:9,
          movie_id:10,
          rating:10,
          created_at:"2019-12-25T20:30:21.606Z",
          updated_at:"2019-12-25T20:30:21.606Z"}
        ]} }
    />);
    wrapper.find('.reset-rating').simulate('click');
    expect(deleteRating).toHaveBeenCalledWith(45, 9);
  });

  it('should invoke fetchRatings when deleteRating is resolved', () => {
    deleteRating(45, 9);
    expect(fetchRatings).toHaveBeenCalledWith(9);
  });

  it('should invoke handleRatingsUpdates on click of any rating button (10)', () => {
   let mockEvent = { target: {value: '2'} };
   wrapper = shallow(<MovieShowPage
     id={2}
     title="Ad Astra"
     poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
     backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
     release_date="2019-09-17"
     overview="The near future, humanity to look to the stars an....."
     average_rating={5.142857142857143}
     isLoggedIn={true}
     updateUser={mockUpdateUser}
     handleError={mockHandleError}
     user={ {id: 9, name: 'Marge', email: 'marge@turing.io', ratings: [
       {id:45,
         user_id:9,
         movie_id:4,
         rating:8,
         created_at:"2019-12-25T20:16:34.893Z",
         updated_at:"2019-12-25T20:16:34.893Z"},
       {id:46,
         user_id:9,
         movie_id:10,
         rating:10,
         created_at:"2019-12-25T20:30:21.606Z",
         updated_at:"2019-12-25T20:30:21.606Z"}
       ]} }
   />);
   wrapper.instance().handleRatingsUpdates = jest.fn();
   wrapper.find('.btn-two').simulate('click', mockEvent);
   expect(wrapper.instance().handleRatingsUpdates).toHaveBeenCalledWith(mockEvent);
  });

  it('should invoke updateRatings fetch when handleRatingsUpdates is called',
    async () => {
      let mockEvent = { target: {value: '2'} };
      await wrapper.instance().handleRatingsUpdates(mockEvent);
      expect(updateRatings).toHaveBeenCalledWith(2, 2, 9);
  });

  it('should invoke fetchRatings when updateRatings resolves', async () => {
    await updateRatings(2, 2, 9);
    expect(fetchRatings).toHaveBeenCalledWith(9);
  });

  // it('should invoke updateUser when fetchRatings resolves after handleRatingsUpdates is invoked', () => {
  //   let mockNewRatings = {id: 9, name: 'Marge', email: 'marge@turing.io',
  //       ratings: [
  //     {id: 45, user_id: 9, movie_id: 8, rating: 8,
  //       created_at: "2019-12-25T20:16:34.893Z",
  //       updated_at: "2019-12-25T20:16:34.893Z"
  //     },
  //     {id: 46, user_id: 9, movie_id: 10, rating: 5,
  //       created_at: "2019-12-25T20:30:21.606Z",
  //       updated_at: "2019-12-25T20:30:21.606Z"
  //     } ]};
  //   updateRatings.mockImplementation(() => {
  //     return Promise.resolve({ rating: { user_id: 9, movie_id: 2, rating: 2 } })
  //   });
  //   fetchRatings.mockImplementation(() => {
  //     return Promise.resolve({ ratings: [
  //       {id: 45, user_id: 9, movie_id: 8, rating: 8,
  //         created_at: "2019-12-25T20:16:34.893Z",
  //         updated_at: "2019-12-25T20:16:34.893Z"
  //       },
  //       {id: 46, user_id: 9, movie_id: 10, rating: 5,
  //         created_at: "2019-12-25T20:30:21.606Z",
  //         updated_at: "2019-12-25T20:30:21.606Z"
  //       } ] })
  //   })
  //   fetchRatings(9);
  //   expect(mockUpdateUser).toHaveBeenCalledWith(mockNewRatings);
  //   // FAILING TEST - NUMBER OF CALLS: 0 - WHY???
  // });

  // HOW DO I MAKE THE TEST BELOW DIFFERENT FROM THE IDENTICAL ONE ABOVE SINCE FETCH RATINGS IS INVOKED MORE THAN ONCE?
  // lines 23-25 are identical to lines 42-44 so maybe we can wrap them in a handler function?
  // it('should invoke updateUser when fetchRatings resolves', () => {
  //
  // });

// FAILING, NUMBER OF CALLS = 0
  // it('should invoke handleError prop if updateRatings rejects', async () => {
  //   let mockEvent = { target: {value: '2'} };
  //
  //   updateRatings.mockImplementation(() => {
  //     return Promise.reject(Error('error '));
  //   });
  //   await wrapper.instance().handleRatingsUpdates(mockEvent);
  //   expect(mockHandleError).toHaveBeenCalledWith('Problem updating your rating, please try again later.');
  // })

  describe('mapsStateToProps', () => {
    it('should return only the necessary information from the redux store',
      () => {
        const mockState = {
          movies: [],
          user: {newUser: 'Tron'},
          isLoggedIn: true,
          errorMessage: '',
          loadingStatus: false,
        };
        const expected = {
          errorMessage: '',
          isLoggedIn: true,
          user: {newUser: 'Tron'}
        };
        const mappedProps = mapStateToProps(mockState);

        expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls dispatch with an updateUser action when updateUser is called',
      () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = updateUser({some: 'thing'});
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.updateUser({some: 'thing'});

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with an handleError action when handleError is called',
      () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = handleError('clever error message');
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.handleError('clever error message');

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
