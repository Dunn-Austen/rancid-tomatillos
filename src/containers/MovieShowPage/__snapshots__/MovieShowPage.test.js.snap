// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieShowPage should match snapshot when logged in and no user rating yet 1`] = `
<div
  className="movie-page"
>
  <img
    alt="Ad Astra"
    className="backdrop"
    src="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
  />
  <h1
    className="movie__title--two"
  >
    Ad Astra
  </h1>
  <p
    className="in-theaters"
  >
    In Theaters: 
    2019-09-17
  </p>
  <p
    className="overview"
  >
    The near future, humanity to look to the stars an.....
  </p>
  <p
    className="average__rating--two"
  >
    Average Rating: 
    5.1
  </p>
  <div
    className="rating-bar"
  >
    <div
      className="rating-scale"
    >
      <button
        className="rating-btn btn-one"
        onClick={[Function]}
        value="1"
      >
        1
      </button>
      <button
        className="rating-btn btn-two"
        onClick={[Function]}
        value="2"
      >
        2
      </button>
      <button
        className="rating-btn btn-three"
        onClick={[Function]}
        value="3"
      >
        3
      </button>
      <button
        className="rating-btn btn-four"
        onClick={[Function]}
        value="4"
      >
        4
      </button>
      <button
        className="rating-btn btn-five"
        onClick={[Function]}
        value="5"
      >
        5
      </button>
      <button
        className="rating-btn btn-six"
        onClick={[Function]}
        value="6"
      >
        6
      </button>
      <button
        className="rating-btn btn-seven"
        onClick={[Function]}
        value="7"
      >
        7
      </button>
      <button
        className="rating-btn btn-eight"
        onClick={[Function]}
        value="8"
      >
        8
      </button>
      <button
        className="rating-btn btn-nine"
        onClick={[Function]}
        value="9"
      >
        9
      </button>
      <button
        className="rating-btn btn-ten"
        onClick={[Function]}
        value="10"
      >
        10
      </button>
    </div>
  </div>
</div>
`;

exports[`MovieShowPage should match snapshot when user is logged in and rating is 10 1`] = `
<div
  className="movie-page"
>
  <img
    alt="Ad Astra"
    className="backdrop"
    src="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
  />
  <h1
    className="movie__title--two"
  >
    Ad Astra
  </h1>
  <p
    className="in-theaters"
  >
    In Theaters: 
    2019-09-17
  </p>
  <p
    className="overview"
  >
    The near future, humanity to look to the stars an.....
  </p>
  <p
    className="average__rating--two"
  >
    Average Rating: 
    5.1
  </p>
  <div
    className="already-rated"
  >
    <p
      className="top-marks"
    >
      My Rating: 
      10
    </p>
    <button
      className="reset-rating"
      onClick={[Function]}
    >
      Reset Rating
    </button>
  </div>
</div>
`;

exports[`MovieShowPage should match snapshot when user is logged in and rating is less than 10 1`] = `
<div
  className="movie-page"
>
  <img
    alt="Ad Astra"
    className="backdrop"
    src="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
  />
  <h1
    className="movie__title--two"
  >
    Ad Astra
  </h1>
  <p
    className="in-theaters"
  >
    In Theaters: 
    2019-09-17
  </p>
  <p
    className="overview"
  >
    The near future, humanity to look to the stars an.....
  </p>
  <p
    className="average__rating--two"
  >
    Average Rating: 
    5.1
  </p>
  <div
    className="already-rated"
  >
    <p
      className="my-rating"
    >
      My Rating: 
      8
    </p>
    <button
      className="reset-rating"
      onClick={[Function]}
    >
      Reset Rating
    </button>
  </div>
</div>
`;

exports[`MovieShowPage should match the MovieShowPage Snapshot with no logged in user 1`] = `
<div
  className="movie-page"
>
  <img
    alt="Ad Astra"
    className="backdrop"
    src="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
  />
  <h1
    className="movie__title--two"
  >
    Ad Astra
  </h1>
  <p
    className="in-theaters"
  >
    In Theaters: 
    2019-09-17
  </p>
  <p
    className="overview"
  >
    The near future, humanity to look to the stars an.....
  </p>
  <p
    className="average__rating--two"
  >
    Average Rating: 
    5.1
  </p>
</div>
`;
